rust/src/data_catalog/storage/mod.rs

```diff
@@ -0,0 +1,194 @@
+//! listing_schema contains a SchemaProvider that scans ObjectStores for tables automatically
+use std::any::Any;
+use std::collections::{HashMap, HashSet};
+use std::path::Path;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use dashmap::DashMap;
+use datafusion::catalog::schema::SchemaProvider;
+use datafusion::datasource::TableProvider;
+use datafusion_common::DataFusionError;
+use futures::TryStreamExt;
+use itertools::Itertools;
+use object_store::ObjectStore;
+
+use crate::storage::config::{configure_store, StorageOptions};
+use crate::{ensure_table_uri, open_table_with_storage_options, DeltaResult};
+
+const DELTA_LOG_FOLDER: &str = "_delta_log";
+
+/// A `SchemaProvider` that scans an `ObjectStore` to automatically discover delta tables.
+///
+/// A subfolder relationship is assumed, i.e. given:
+/// authority = s3://host.example.com:3000
+/// path = /data/tpch
+///
+/// A table called "customer" will be registered for the folder:
+/// s3://host.example.com:3000/data/tpch/customer
+///
+/// assuming it contains valid deltalake data, i.e a `_delta_log` folder:
+/// s3://host.example.com:3000/data/tpch/customer/_delta_log/
+pub struct ListingSchemaProvider {
+    authority: String,
+    path: object_store::path::Path,
+    /// Underlying object store
+    store: Arc<dyn ObjectStore>,
+    /// A map of table names to a fully quilfied storage location
+    tables: DashMap<String, String>,
```

Comment:
Is there an artifact of implementing this `SchemaProvider` that requires the use of DashMap? Since this adds a new dependency, I'm curious what forces this requirement since it's not clear to me

Reply:
We could also go the "classic" route using `Arc<Mutex<...>>`, however since datafusion internally switched in most places, we already have that as a transient dependency if we use datafusion.

It does however simplify all code where we use this.